Git-commit: 9a84818ec31d585c1a236ec4657a63e77ed9bc08
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Fri, 26 Feb 2021 20:21:32 +0800
Subject: [PATCH] arm64/mpam: get num_partids from system regs instead of hard
 code
Patch-mainline: Queued in openEuler repo, version 5.10.0-4.5.0
References: openEuler-21.03
Git-repo: https://gitee.com/openeuler/kernel.git

hulk inclusion
category: bugfix
bugzilla: 48265
CVE: NA

--------------------------------

Previously hard coded for num of partids, now get them
through the system registers.

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Kai Liu <kai.liu@suse.com>
---
 arch/arm64/kernel/mpam.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/kernel/mpam.c b/arch/arm64/kernel/mpam.c
index dae2303b2449..44ba7ea4fda3 100644
--- a/arch/arm64/kernel/mpam.c
+++ b/arch/arm64/kernel/mpam.c
@@ -322,9 +322,17 @@ static int closid_free_map;
 
 void closid_init(void)
 {
-	int resctrl_min_closid = 32;
+	struct resctrl_resource *r;
+	struct raw_resctrl_resource *rr;
+	int num_closid = INT_MAX;
 
-	closid_free_map = BIT_MASK(resctrl_min_closid) - 1;
+	for_each_resctrl_resource(r) {
+		if (r->alloc_enabled) {
+			rr = r->res;
+			num_closid = min(num_closid, rr->num_partid);
+		}
+	}
+	closid_free_map = BIT_MASK(num_closid) - 1;
 
 	/* CLOSID 0 is always reserved for the default group */
 	closid_free_map &= ~1;
-- 
2.31.1

